Impro1044 part 2: split merge_cubes functionality

cgsandford opened the pull request:
Refactored merge_cubes into two classes.  Added new unit tests and migrated existing tests from moved / refactored utilities.

Partially addresses #835 

Description of changes:

- Moved  `_check_coord_bounds`, `_check_bounds_ranges`, `_equalise_cell_methods` and `_equalise_cubes` into new `MergeCubes` class
- Refactored `MergeCubes._equalise_cubes` to use the general utility `equalise_cube_attributes`, and removed `_equalise_cube_attributes` (which also created model ID and configuration coordinates)
- Created class `MergeCubesForWeightedBlending` with a new function to create model ID and configuration coordinates, and calling existing function `rationalise_blend_time_coords` (this will replace the calls in the improver-weighted-blending CLI).  This sets up cubes for blending and then calls the new `MergeCubes.process` method.
- Moved model blending test cases out of test_merge_cubes.py and into MergeCubesForWeightedBlending.py.  (Test setup has been refactored but test cases / coverage remains the same.)  These demonstrate that reordering the setup of cubes for blending has not affected the outcome after merging.


Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

cgsandford commented:
Note: this overlaps with #838.  Once that has been merged I'll need to sort out the "threshold" coordinate issue in `MergeCubes`.

gavinevans commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -314,363 +316,383 @@ def concatenate_cubes(
+        # if coord_mismatch_error_keys includes "threshold", replace entry with
+        # standard name of threshold-type coordinate on input cubes
+        if "threshold" in self.coord_mismatch_error_keys:
+            try:
```
I think this is a good use of try/except/else. I think that perhaps there are other places in the codebase, where we should be using try/except/else, rather than just try/except, as this formulation actually makes it clear what is being "tried".

cgsandford commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -312,363 +314,372 @@ def concatenate_cubes(
+        cubelist = self._equalise_cube_coords(cubelist)
+        cubelist = self._equalise_cell_methods(cubelist)
+        for cube in cubelist:
+            check_cube_not_float64(cube, fix=True)  # TODO why here?
```
Yes, I wasn't envisaging changing this at the moment, but I think it's worth leaving the TODO as we we will want to fix this - it's a bad idea to have a type cast buried where it might cause problems and be difficult to find.  This probably links in with the metadata and datatype conformance work.

cgsandford commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -312,363 +314,372 @@ def concatenate_cubes(
+        # check for unmatching coords (returns a list of dictionaries, each
+        # with an entry for each mismatched coord.  If all coords match,
+        # returns a list of empty dictionaries.)
+        unmatching_coords = compare_coords(cubes)
```
I wonder whether `compare_coords` would benefit from the same attention as `compare_attributes`, perhaps as part of the same Jira follow-on ticket?

gavinevans replied:
Yes, potentially a follow-on ticket to think about what the `compare_coords` function is returning and whether this can be rationalised, would sound reasonable.

gavinevans replied:
Would you be happy to create a ticket for this?

cgsandford commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -312,363 +314,372 @@ def concatenate_cubes(
+                        match_coord = later_cube.coord(coord)
+                    except CoordinateNotFoundError:
+                        continue
+                    if coord.bounds is None and match_coord.bounds is None:
```
:+1:

cgsandford commented:
@gavinevans Thanks for your comments, I've addressed the easy ones I think (docstrings etc).

On the MergeCubesForWeightedBlending plugin and tests, it looks like I haven't split that very well between `__init__` and `process`, and your comments have highlighted some checks I think I've missed altogether.  So I'll need to spend a little time reworking that next week before it's ready to re-review.  Thanks for picking up on those details.

cgsandford commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -312,363 +314,372 @@ def concatenate_cubes(
+            cubes (iris.cube.CubeList or iris.cube.Cube):
+                Cubes to be merged.
+            blend_coord (str):
+                Name of coordinate over which blending will be performed.
```
I don't think we're restricted to supporting only these three coordinates.  It's just that blend time rationalisation only has an effect on the input cubes if one of these is the blend coordinate.  I don't think the class will fail if given any other coordinate, eg realisation, to blend over.  I'll add a test to show this.

cgsandford commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -312,363 +314,372 @@ def concatenate_cubes(
+        Args:
+            cubes (iris.cube.CubeList or iris.cube.Cube):
+                Cubes to be merged.
+            blend_coord (str):
```
I've moved all of them except `cycletime`, as I can see an argument for keeping this with the `process` method.

cgsandford commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -312,363 +314,372 @@ def concatenate_cubes(
+            return cubes_in
+
+        # test for model blend
+        if "model" in blend_coord and self.model_id_attr is None:
```
This was deliberate.  I rather like the idea of this class being opaque - the user can specify "model" as the blend coordinate, and the specific names / nature of the `model_id` and `model_configuration` coordinate is known only within the blending code.  This gives us flexibility to change how we handle this metadata without affecting the user interface.  We can also then get the if / else statement out of the CLI and stop worrying about if the CLI argument is `model_id`, `model_configuration` or just `model`.

What do you think?

cgsandford replied:
(Note the use of `if "model" in blend_coord` is also consistent with what we're doing in `rationalise_blend_time_coords`.)

gavinevans replied:
Do you think that it is worth adding something to the `blend_coord` docstring (`Name of coordinate over which blending will be performed.`)? Something like `Name of coordinate over which blending will be performed. If blending models, then a 'model' coordinate is supported.` This is just because `model` is not actually the name of any coordinate, despite the docstring saying that `blend_coord` should be a `name of a coordinate`.

cgsandford replied:
Updated docstring to explain this.

cgsandford commented on lib/improver/tests/utilities/cube_manipulation_new/test_MergeCubesForWeightedBlending.py:
```diff
@@ -0,0 +1,251 @@
+    def test_basic(self):
+        """Test single cube is returned unmodified"""
+        cube = self.cube_enuk.copy()
+        result = self.plugin.process(cube, "model")
```
As I've noted above, I'd quite like the interface here not to require the user to know about the internal workings of the class.  So `model`, `model_id` or `model_configuration` should be equally acceptable as inputs here.  This is consistent with the design of `rationalise_blend_time_coords` as well.

cgsandford commented on lib/improver/tests/utilities/cube_manipulation_new/test_MergeCubesForWeightedBlending.py:
```diff
@@ -0,0 +1,251 @@
+        msg = "Cube with mismatching time bounds ranges cannot be blended"
+        with self.assertRaisesRegex(ValueError, msg):
+            MergeCubesForWeightedBlending().process(
+                [self.cube_ukv, cube2], "forecast_reference_time")
```
Added.

cgsandford commented:
@gavinevans I think I've now addressed your comments and have reworked the split between `__init__` and `process` to check for things more sensibly (specifically we don't want to create model ID coordinates unless we're blending over models!).  So this should now be ready to re-review.

gavinevans commented on lib/improver/tests/utilities/cube_manipulation_new/test_MergeCubes.py:
```diff
@@ -0,0 +1,384 @@
+        self.plugin._check_time_bounds_ranges(self.matched_cube)
+
+
+class Test_process(IrisTest):
```
A test to cover the `check_time_bounds_ranges` keyword argument would be useful. What do you think?

cgsandford replied:
Added (pretty much copied from test_MergeCubesForWeightedBlending.py)

gavinevans commented:
@cgsandford Thanks for these changes 👍  
I've now finished my re-review and just made a couple of very minor comments.

fionaRust commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -314,367 +316,417 @@ def concatenate_cubes(
+    Accounts for differences in attributes and coordinates to avoid merge
+    failures and anonymous dimensions.
+    """
+    def __init__(self):
```
Why are these not passed in? What is the rational for silently ignoring these and the coord_mismatch_error_keys as well?

cgsandford commented:
Quick comment that answers some of your questions / comments above.

With this refactor, I haven't actually changed anything about what functionality is called and when.  I've simply moved out the functionality from where "merging=True" (or whatever it was to indicate blending - I honestly can't remember!) into a separate class.  So I'm pretty sure you're right, this won't be the correct split, it's just highlighted exactly how we have been doing it in the past.  This also applies to your question "why aren't we passing these in": they were hard-coded in the previous code, so I've kept them hard-coded here (time-lagging does use concatenation, as realization is a dimension coordinate).  In this way I've done my best to ensure this code behaves (almost) exactly how the old code behaved.

My suggestion is that we don't change anything about the split until the next PR, which will be when I replace the old code.  At that point, I can work against the CLI tests to figure out whether we're using any of the pathways you've highlighted we might want to change or remove.  I can also change the calls to pass in lists of unmatching attributes etc as required.  But if I make any functional changes before that I'm more likely to encounter problems with CLI tests failing because of something we haven't anticipated.

What do you think?

cgsandford commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -314,367 +316,417 @@ def concatenate_cubes(
+            # throw an error for specific coordinate mismatches
+            for error_key in self.coord_mismatch_error_keys:
+                # if necessary, translate threshold coordinate name
+                if error_key == "threshold":
```
The new naming convention means that threshold is no longer called "threshold", but will have (for example) `standard_name="air_temperature"` and `var_name="threshold"`.  Since the mismatched keys will contain an actual coordinate standard or long name, we have to check the cubes for what the "threshold-type" coordinate is actually called.  This is one of the things incorporated from #846.

fionaRust replied:
How does this overlap with the change to the `self.coord_mismatch_error_keys` in the process method? If I'm reading the code right we'll never get into this loop because "threshold" will have been replaced in the `self.coord_mismatch_error_keys` the process method.

cgsandford replied:
This is an excellent point.  I will revert the behaviour in `_equalise_cube_coords` as I think that was what I originally intended to do.

fionaRust commented:
> Quick comment that answers some of your questions / comments above.
> 
> With this refactor, I haven't actually changed anything about what functionality is called and when. I've simply moved out the functionality from where "merging=True" (or whatever it was to indicate blending - I honestly can't remember!) into a separate class. So I'm pretty sure you're right, this won't be the correct split (or more precisely, I think we'll be able to get rid of a fair bit of the `MergeCubes` class), it's just highlighted exactly how we have been doing it in the past. This also applies to your question "why aren't we passing these in": they were hard-coded in the previous code, so I've kept them hard-coded here (time-lagging does use concatenation, as realization is a dimension coordinate). In this way I've done my best to ensure this code behaves (almost) exactly how the old code behaved.
> 
> My suggestion is that we don't change anything about the split until the next PR, which will be when I replace the old code. At that point, I can work against the CLI tests to figure out whether we're using any of the pathways you've highlighted we might want to change or remove. I can also change the calls to pass in lists of unmatching attributes etc as required. But if I make any functional changes before that I'm more likely to encounter problems with CLI tests failing because of something we haven't anticipated.
> 
> What do you think?


I'm happy with this approach and I'll finish reviewing this with this approach in mind. I'm happy for you to do this however you think will give you less problems overall

fionaRust commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -314,367 +316,417 @@ def concatenate_cubes(
+                Merged cube.
+        """
+        # if input is already a single cube, return unchanged
+        if isinstance(cubes_in, iris.cube.Cube):
```
or if len(cubelist) ==1? (relates to a comment somewhere else about unit tests for cubelists with only one cube in)

cgsandford replied:
Sorted (see above)

fionaRust commented on lib/improver/tests/utilities/cube_manipulation_new/test_MergeCubes.py:
```diff
@@ -0,0 +1,406 @@
+        self.plugin._check_time_bounds_ranges(self.matched_cube)
+
+
+class Test_process(IrisTest):
```
I think you need unit tests for:
* a single cube in rather than a cubelist
* that your input cubes are unchanged
* That the self.coord_mismatch_error_keys are updated appropriately for a threshold coord

cgsandford replied:
OK the last of these took a little reworking, but I've now added tests to cover this.

fionaRust commented:
@cgsandford  I think I've finished reviewing this now

cgsandford commented on lib/improver/tests/utilities/cube_manipulation_new/test_MergeCubes.py:
```diff
@@ -0,0 +1,406 @@
+                                 ["threshold"])
+
+
+class Test__equalise_cubes(IrisTest):
```
`_equalise_cubes` is an odd function because it just calls a series of functions that are unit tested elsewhere in the module.  So I didn't think too many checks were needed, and I chose the aspects for which input cubes were easiest to set up (`_equalise_coords` requires a particularly convoluted setup).  I should cover the float64 aspect though.

fionaRust replied:
I don't think you've done the float64 unit test yet.

cgsandford replied:
Now added a couple of datatype tests covering demotion of float64 and non-promotion of integer types.

cgsandford commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -314,367 +316,417 @@ def concatenate_cubes(
-        cubelist[i] = iris.util.squeeze(cube)
+    def _equalise_cube_coords(self, cubes):
+        """
+        Function to equalise coordinates that do not match.
```
I've added this docstring after a bit of thought, which unfortunately raises some questions as to whether what this function is doing is sane (it's probably not).  Suggest though that I address that in the next PR.

fionaRust replied:
Ok, that's fine by me.

cgsandford commented on lib/improver/tests/utilities/cube_manipulation_new/test_MergeCubesForWeightedBlending.py:
```diff
@@ -0,0 +1,308 @@
+        self.assertIsNone(plugin.model_id_attr)
+
+
+class Test__create_model_coordinates(IrisTest):
```
The model ID attribute isn't removed, but is renamed to "blend".  I check this in `test_basic`.  I've updated the docstring.

I've now added a test for values and one to test what happens when I pass in two cubes with the same model ID attribute.  That currently doesn't fail, but the model configuration (auxiliary) coordinate obviously contains duplicate points.  I wonder if we want to throw an error under those circumstances?  I think maybe we should.

fionaRust replied:
I think we probably should raise an error too.

cgsandford replied:
Added.

cgsandford commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -314,367 +316,417 @@ def concatenate_cubes(
+            ValueError:
+                If self.model_id_attr is not present on all cubes
+        """
+        if self.model_id_attr is None:
```
You're right that's tidier.  I've moved that into `process`/

cgsandford commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -314,367 +316,417 @@ def concatenate_cubes(
+            cubelist.append(cube.copy())
+
+        # TODO move rationalise_blend_time_coords into this class
+        rationalise_blend_time_coords(
```
Yes, that's the plan.  It makes more sense to call from a plugin.

cgsandford commented on lib/improver/utilities/cube_manipulation_new.py:
```diff
@@ -314,367 +316,417 @@ def concatenate_cubes(
-            else:
-                cubelist.append(cube)
+# TODO move this class into improver.blending.weighted_blend
+class MergeCubesForWeightedBlending():
```
Agreed.

cgsandford commented on lib/improver/tests/utilities/cube_manipulation_new/test_MergeCubesForWeightedBlending.py:
```diff
@@ -0,0 +1,308 @@
+            self.plugin._create_model_coordinates(self.cubelist)
+
+
+class Test_process(IrisTest):
```
If it helps.  My thought process is that there are three use cases for WeightedBlendAcrossWholeDimension:

- Cycle blending (over forecast period)
- Multi-model blending (over model_id or model_configuration)
- Collapsing realizations (via `improver-threshold` with the  `--collapse_coord` option)

So I've use those three blend coordinates in the "process" unit tests.  Collapsing realizations doesn't require any special setup, so that's just the one test to make sure `MergeCubes` is called without modifying the inputs.  For model blending I've tested model coordinate creation; and for both cycle and model blending I've tested `rationalise_blend_time_coordinates` indirectly, by setting up test cubes with different forecast periods and checking that the merged output has scalar `time` and `forecast_reference_time coordinates`.

fionaRust replied:
Thank you for explaining. I'm happy with this test coverage.

cgsandford commented on lib/improver/tests/utilities/cube_manipulation_new/test_MergeCubes.py:
```diff
@@ -0,0 +1,406 @@
+            self.plugin._check_dim_coord_bounds(self.cubelist)
+
+
+class Test__equalise_cell_methods(IrisTest):
```
Doesn't happen.  I've set up both `process` methods to return immediately if the input is a single cube.  I should probably update these statements to return immediately from a single-item cubelist, too.

cgsandford replied:
And some tests...

cgsandford commented:
@fionaRust I think I've now addressed all your comments, mostly without question!

Regarding the TODOs, yes, I'm expecting to address these as part of the ticket / issue #835 .  I'm anticipating two more PRs: one to migrate everything to the new plugins and fix any bugs; and then one to move the blending preparation into the weighted blending module.
